pkg/credentialprovider/external.go

```diff
@@ -0,0 +1,234 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package credentialprovider
+
+import (
+	"bytes"
+	"fmt"
+	"net/url"
+	"os/exec"
+	"sort"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/klog"
+	"k8s.io/kubernetes/pkg/credentialprovider/apis/registrycredentials"
+	"k8s.io/kubernetes/pkg/credentialprovider/apis/registrycredentials/v1alpha1"
+)
+
+var (
+	scheme = runtime.NewScheme()
+	codecs = serializer.NewCodecFactory(scheme)
+)
+
+func init() {
+	registrycredentials.AddToScheme(scheme)
+	v1alpha1.AddToScheme(scheme)
+}
+
+type externalProviderKeyring struct {
+	// map of registry matcher to provider (each provider may have multiple)
+	providers map[string]ExternalCredentialProvider
+	// reverse sorted index of registries
+	index []string
+}
+
+func NewExternalProviderKeyring(configPath string) (DockerKeyring, error) {
+	keyring := &externalProviderKeyring{
+		providers: make(map[string]ExternalCredentialProvider),
+		index:     make([]string, 0),
+	}
+
+	config, err := readExternalProviderConfig(configPath)
+	if err != nil {
+		return keyring, err
+	}
+
+	for _, p := range config.Providers {
+		provider := externalCredentialProvider{
+			command: p.Exec.Command,
+			args:    p.Exec.Args,
+			env:     convertEnvs(p.Exec.Env),
+		}
+		for _, m := range p.ImageMatchers {
+
+			value := m
+			if !strings.HasPrefix(value, "https://") && !strings.HasPrefix(value, "http://") {
```

Comment:
Why are you accepting https:// prefixes?

Reply:
This was previous behavior--if its wrong I'll fix it but I'm struggling to find references one way or the other.

Reply:
This is the commit it was originally added in the existing provider: https://github.com/kubernetes/kubernetes/commit/8c2731491532

Reply:
>  Unqualified host:port pairs are valid Docker auth fields
>
> The dockercfg and .docker/config.json files can contain host:path
combos, which are not correctly parsed by the keyring.

Reply:
An image reference can't contain a url scheme:

https://github.com/docker/distribution/blob/master/reference/reference.go#L4-L24

The referenced change seems like a hack. Do we actually want to rebase it or fix the problem?

Reply:
Given that this is alpha, my recommendation is to add the base functionality you need and add functionality as feature requests come up.

Reply:
+1